lib/elsa/group/acknowledger.ex

```diff
@@ -0,0 +1,131 @@
+defmodule Elsa.Group.Acknowledger do
+  @moduledoc """
+  Handles acknowledgement of messages to the
+  group coordinator to prevent the group manager
+  from queuing up messages for acknowledgement
+  when events such as a rebalance occur.
+  """
+  use GenServer
+  require Logger
+  import Elsa.Supervisor, only: [registry: 1]
+
+  @doc """
+  Trigger acknowledgement of processed messages back to the cluster.
+  """
+  @spec ack(String.t(), Elsa.topic(), Elsa.partition(), Elsa.Group.Manager.generation_id(), integer()) :: :ok
+  def ack(connection, topic, partition, generation_id, offset) do
+    acknowledger = {:via, Elsa.Registry, {registry(connection), __MODULE__}}
+    GenServer.cast(acknowledger, {:ack, topic, partition, generation_id, offset})
+  end
+
+  @doc """
+  Trigger acknowledgement of processed messages back to the cluster.
+  """
+  @spec ack(String.t(), %{
+          topic: Elsa.topic(),
+          partition: Elsa.partition(),
+          generation_id: Elsa.Group.Manager.generation_id(),
+          offset: integer()
+        }) :: :ok
+  def ack(connection, %{topic: topic, partition: partition, generation_id: generation_id, offset: offset}) do
+    ack(connection, topic, partition, generation_id, offset)
+  end
+
+  @doc """
+  Sync the group generation ID back to the acknowledger state for validation.
+  """
+  @spec update_generation_id(pid() | atom(), Elsa.Group.Manager.generation_id()) :: :ok
+  def update_generation_id(acknowledger, generation_id) do
+    GenServer.cast(acknowledger, {:update_generation, generation_id})
+  end
+
+  @doc """
+  Retrieve the latest offset for a topic and partition. Primarily used for reinitializing
+  consumer workers to the latest unacknowledged offset after a rebalance or other disruption.
+  """
+  @spec get_latest_offset(pid() | atom(), Elsa.topic(), Elsa.partition()) :: Elsa.Group.Manager.begin_offset()
+  def get_latest_offset(acknowledger, topic, partition) do
+    GenServer.call(acknowledger, {:get_latest_offset, topic, partition})
+  end
+
+  @doc """
+  Instantiate an acknowledger process and register it to the Elsa registry.
+  """
+  @spec start_link(term()) :: GenServer.on_start()
+  def start_link(opts) do
+    connection = Keyword.fetch!(opts, :connection)
+    GenServer.start_link(__MODULE__, opts, name: {:via, Elsa.Registry, {registry(connection), __MODULE__}})
+  end
+
+  @impl GenServer
+  def init(opts) do
+    connection = Keyword.fetch!(opts, :connection)
+
+    state = %{
+      connection: connection,
+      current_offsets: %{},
+      generation_id: nil,
+      group_coordinator_pid: nil
+    }
+
+    {:ok, state, {:continue, :get_coordinator}}
+  end
+
+  @impl GenServer
+  def handle_continue(:get_coordinator, state) do
+    retriever = fn -> Elsa.Registry.whereis_name({registry(state.connection), :brod_group_coordinator}) end
+    group_coordinator_pid = retrieve_coordinator(retriever, 5)
```

Comment:
What if we move the brod_group_coordinator under the Elsa.Group.Supervisor (wrapped by Elsa.Wrapper)?  That should remove the need for the retrieve_coordinator loop and simplify the death handling login in Elsa.Group.Manager as well.

Reply:
I’ll give it a try and see how it goes.

Reply:
Turns out it works.